cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -174,42 +150,85 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte, clustername
-				close(stopChan)
-			})
-		}, constants.DiscoveryRetryInterval, stopChan)
-	}()
-
-	select {
-	case <-time.After(discoveryTimeout):
-		once.Do(func() {
-			close(stopChan)
-		})
-		err := errors.Errorf("abort connecting to API servers after timeout of %v", discoveryTimeout)
-		klog.V(1).Infof("[discovery] %v\n", err)
-		wg.Wait()
+// validateClusterInfoToken validates that the JWS token present in the cluster info ConfigMap is valid
+func validateClusterInfoToken(insecureClusterInfo *v1.ConfigMap, token *kubeadmapi.BootstrapTokenString) ([]byte, error) {
+	insecureKubeconfigString, ok := insecureClusterInfo.Data[bootstrapapi.KubeConfigKey]
+	if !ok || len(insecureKubeconfigString) == 0 {
+		return nil, errors.Errorf("there is no %s key in the %s ConfigMap. This API Server isn't set up for token bootstrapping, can't connect",
+			bootstrapapi.KubeConfigKey, bootstrapapi.ConfigMapClusterInfo)
+	}
+
+	detachedJWSToken, ok := insecureClusterInfo.Data[bootstrapapi.JWSSignatureKeyPrefix+token.ID]
+	if !ok || len(detachedJWSToken) == 0 {
+		return nil, errors.Errorf("token id %q is invalid for this cluster or it has expired. Use \"kubeadm token create\" on the control-plane node to create a new valid token", token.ID)
+	}
+
+	if !bootstrap.DetachedTokenIsValid(detachedJWSToken, insecureKubeconfigString, token.ID, token.Secret) {
+		return nil, errors.New("failed to verify JWS signature of received cluster info object, can't trust this API Server")
+	}
+
+	return []byte(insecureKubeconfigString), nil
+}
+
+// validateClusterCA validates the cluster CA found in the insecure kubeconfig
+func validateClusterCA(insecureConfig *clientcmdapi.Config, pubKeyPins *pubkeypin.Set) ([]byte, error) {
+	var clusterCABytes []byte
+	for _, cluster := range insecureConfig.Clusters {
+		clusterCABytes = cluster.CertificateAuthorityData
+	}
+
+	clusterCAs, err := certutil.ParseCertsPEM(clusterCABytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse cluster CA from the %s ConfigMap", bootstrapapi.ConfigMapClusterInfo)
+	}
+
+	// Validate the cluster CA public key against the pinned set
+	err = pubKeyPins.CheckAny(clusterCAs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cluster CA found in %s ConfigMap is invalid", bootstrapapi.ConfigMapClusterInfo)
+	}
+
+	return clusterCABytes, nil
+}
+
+// getClusterInfo creates a client from the given kubeconfig if the given client is nil,
+// and requests the cluster info ConfigMap using PollImmediate.
+// If a client is provided it will be used instead.
+func getClusterInfo(client clientset.Interface, kubeconfig *clientcmdapi.Config, token *kubeadmapi.BootstrapTokenString, interval, duration time.Duration) (*v1.ConfigMap, error) {
+	var cm *v1.ConfigMap
+	var err error
+
+	// Create client from kubeconfig
+	if client == nil {
+		client, err = kubeconfigutil.ToClientSet(kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), duration)
```

Comment:
The context here gives me an idea. What about creating a root context (at the beginning of `retrieveValidatedConfigInfo`), that contains all the timeout. Then we can pass it down to this func. This local context can then be created like so:
```golang
ctx, cancel := context.WithCancel(parentCtx)
```

This has the added benefit of not having to deal with timeout calculation between all leaf `JitterUntil` calls. The local context is still needed to stop `JitterUntil` from spinning.

Reply:
my initial impression is that this is not needed, but i will double check it.

Reply:
> This has the added benefit of not having to deal with timeout calculation between all leaf JitterUntil calls. The local context is still needed to stop JitterUntil from spinning.

unless i'm wrong this is problematic. we want a new context with timeout calculated each time getClusterInfo() is called.

https://golang.org/src/context/context.go?s=14448:14525#L453
```
func WithTimeout(parent Context, timeout time.Duration) (Context, CancelFunc) {
	return WithDeadline(parent, time.Now().Add(timeout))
```

i'd prefer we keep the code the way it is.

for the current use of JitterUntil we technically don't need the function to accept a channel, but a "duration" argument like Poll* would have been fine too.

Reply:
Ok, let's keep things like this.
/lgtm